stdlib/public/core/UnicodeScalar.swift

```diff
@@ -428,9 +428,10 @@ extension Unicode.Scalar.UTF16View : RandomAccessCollection {
   ///   `endIndex` property.
   @inlinable
   public subscript(position: Int) -> UTF16.CodeUnit {
-    return position == 0 ? (
-      endIndex == 1 ? UTF16.CodeUnit(value.value) : UTF16.leadSurrogate(value)
-    ) : UTF16.trailSurrogate(value)
+    _internalInvariant((0..<self.count).contains(position))
```

Comment:
This change just made the control flow logic easier to understand. I can also not touch this declaration in this PR. We can trap in a future PR.

Reply:
This is a functional change. With the new `_internalInvariant`, this no longer allows subscripting with arbitrary indices for non-BMP runes in debug stdlibs, while the original code did allow that.

Reply:
It's adding an assert, because it would be a bug if the stdlib does so. I don't think it's pragmatic to consider all assertions as being "functional changes" to the standard library.

Reply:
AFAICT, the stdlib does indeed call this with unvalidated transcoded offsets:

https://github.com/apple/swift/pull/23834/files#diff-610243ed72b2ff2c91619cdc75fea5b1R254

Reply:
Right, which is exactly why an assertion is useful. I don't understand, what is your point?

Reply:
My point is that `_internalInvariant` is not the correct way to check for coding errors outside of the stdlib. 

If you intend to trap if `String.UTF16View.subscript` is given an index with some nonsensical transcodedOffset, then we should add a precondition checking for that, either here or there.

If you intend to continue allowing such indices there, then this check needs to be removed.

I don't think it is desirable for user code to generate `_internalInvariant` failures in Debug stdlibs.

Reply:
> My point is that _internalInvariant is not the correct way to check for coding errors outside of the stdlib.

I agree. I added this assertion to detect unexpected inputs coming from the stdlib itself. I could remove this assertion from this PR, but it would degrade the quality of our testing.

> If you intend to trap if String.UTF16View.subscript is given an index with some nonsensical transcodedOffset, then we should add a precondition checking for that, either here or there.

For this PR, I do not intend to change the semantics or behavior of this public API from user code. That does sound like a good thing to add, though, and so we'd have two methods here:

public `subscript` would check and trap, and `subscriptImpl` (or whatever it's named) would do the `_internalInvariant`. The stdlib's code would call the latter.

Would you like to open a SR for that change?

> I don't think it is desirable for user code to generate _internalInvariant failures in Debug stdlibs.

I do not understand this position, could you clarify? This is true of any assertions, are you arguing the stdlib shouldn't have assertions? Why is user code running with a assertions-enabled stdlib relevant to this discussion?

Reply:
> For this PR, I do not intend to change the semantics or behavior of this public API from user code.

I'm trying to point out that you did! The addition of that `_internalInvariant` is a strong value judgement on the validity of the expression `("😞" as UnicodeScalar).utf16[9000]`.

My position is that it's not okay to let this expression run "successfully" in release builds of the stdlib, but to make it hit an assert in debug builds. This is not a valid use of `_internalInvariant`.

Reply:
> My position is that it's not okay to let this expression run "successfully" in release builds of the stdlib, but to make it hit an assert in debug builds. This is not a valid use of _internalInvariant.

That's all uses of `_internalInvariant`. If we had a compile-time proof, we wouldn't need `_internalInvariant`

Reply:
In a debug build of the stdlib, a _user_ error without any bugs in the stdlib itself should not trigger an `_internalInvariant`.

I think what @lorentey is saying is that `_internalInvariant` should not be used to diagnose errors on the part of the _caller_ of a public API--even when the intention is to use this for internal diagnostics. The diagnostics cannot distinguish between calls by the stdlib itself and calls by an end user, and errors on the part of the latter cannot by themselves be in violation of an _internal_ invariant.